### PR TITLE
Probe CrossHair capability instead of gating on Python 3.15 (#109)

### LINKED ISSUE #109: Track CrossHair support for Python 3.15 CALL_KW opcode
## Summary

Python 3.15 introduces bytecode that the current `crosshair-tool` tracer does
not support. Importing the optional `crosshair.hypothesis` integration from
`cuprum/unittests/test_build_final_results_property.py` fails during pytest
collection before the property test can run.

## Failure

On Python 3.15.0b2, `make test` reports:

```text
crosshair.tracers.TraceException: The C-level tracer does not support these opcodes: CALL_KW
```

The error is raised while `importlib.util.find_spec("crosshair.hypothesis")`
imports CrossHair internals.

## Expected outcome

Track CrossHair support for Python 3.15 and remove the temporary skip once the
installed CrossHair release supports the new opcode set.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch replaces the hard-coded Python 3.15 skip for the CrossHair line-splitting contracts with a runtime capability probe, so the gate self-resolves once CrossHair supports the interpreter.

Closes [#109](https://github.com/leynos/cuprum/issues/109).

`test_line_splitting.py` gated on `sys.version_info[:2] == (3, 15)` to avoid importing CrossHair on Python 3.15, where an older `crosshair-tool` could not trace the `CALL_KW` opcode and raised `crosshair.tracers.TraceException` during import. The branch instead probes CrossHair at import time and degrades to skipping the symbolic checks only for expected unavailability cases: a missing dev dependency (`ImportError`) or an interpreter whose opcode set CrossHair cannot yet trace (`TraceException`). Unexpected probe failures are re-raised immediately, and expected fallback paths emit a warning that records the exception type and message. The installed `crosshair-tool` 0.0.104 already registers a `CALL_KW` handler and reports the opcode as supported, so the probe succeeds on the supported interpreters; no manual version bump is needed when CrossHair catches up to a new Python.

## Review walkthrough

- [cuprum/unittests/test_line_splitting.py](https://github.com/leynos/cuprum/blob/issue-109-crosshair-capability-probe/cuprum/unittests/test_line_splitting.py): the `_IS_PYTHON_315` version gate (and the now-unused `import sys`) are removed in favour of the documented capability probe. Review feedback narrows the fallback to `ImportError` and `TraceException`, warns on expected degradation, and tests fallback bindings directly.
- [docs/developers-guide.md](https://github.com/leynos/cuprum/blob/issue-109-crosshair-capability-probe/docs/developers-guide.md): the skip rationale is updated to describe the probe.

## Validation

- `make check-fmt`: pass
- `make lint`: pass
- `make typecheck`: pass
- `make test`: pass (593 passed, 44 skipped; the line-splitting CrossHair contracts pass with CrossHair running)
- targeted `uv run pytest -q cuprum/unittests/test_line_splitting.py`: pass (16 passed)
- `make markdownlint`: pass from the previous revision; no Markdown files changed in the review-feedback commit
- `coderabbit review --agent`: 0 findings from the previous revision

## Notes

I cannot empirically verify the probe on a Python 3.15 interpreter from this environment (it runs 3.14), but the probe is interpreter-agnostic by construction: it runs CrossHair when CrossHair can trace the interpreter and skips otherwise. The sibling maturin-build test retains its own separate 3.15 skip (maturin support), which is out of scope here.

## Summary by Sourcery

Probe CrossHair capability at test import time instead of hard-coding a Python 3.15 version gate for line-splitting contracts, and update related documentation and tests accordingly.

Documentation:

- Document CrossHair capability probing in the developers guide instead of a fixed Python 3.15 skip.

Tests:

- Probe CrossHair availability via a narrowed fallback in line-splitting tests, increase CrossHair contract confirmation timeout, and add clarifying docstrings around concurrency test lock helpers.
- Cover expected CrossHair fallback diagnostics and disabled symbolic-check bindings for `ImportError` and `TraceException` probe failures.

## References

- Lody session: https://lody.ai/leynos/sessions/a9ad6fbe-a489-4341-9a1e-ccbb755ffbe7

### CODE REVIEW

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai resume

File: cuprum/unittests/test_line_splitting.py [hunk 1]
```diff
@@ -30,24 +29,32 @@
...
+    from crosshair.statespace import MessageType
+    from crosshair.test_util import check_states
+except BaseException as _crosshair_exc:
+    # ``crosshair.tracers.TraceException`` subclasses ``BaseException`` (not
+    # ``Exception``) and is raised while importing the tracer module itself,
+    # so it cannot be named here without re-triggering the failing import.
+    # Re-raise genuine control-flow exceptions and treat anything else
+    # (ImportError, TraceException) as "CrossHair unavailable".
+    if isinstance(_crosshair_exc, KeyboardInterrupt | SystemExit | GeneratorExit):
+        raise
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_line_splitting.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat documentation and validation coverage as in scope).

### ❌ Failed checks (4 errors, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests verify helper functions in isolation but lack integration coverage for critical behaviours: module-level symbol binding, BaseException vs Exception catching, and control-flow exception re-rai... | Add integration tests verifying: (1) check\_states/symbols are None when probe fails; (2) symbolic test skips using check\_states; (3) control-flow exceptions (KeyboardInterrupt, SystemExit) are re-raised by probe failure handler. |
| Unit Architecture | ❌ Error | Module-level import executes side-effects (warnings.warn) through `_crosshair_unavailable_bindings`, whose name suggests data transformation only. Side-effects must be visible in names and handled... | Separate side-effects: create a pure query function (returning reason string) and a distinct command function (executing the warning). Call the command at module level explicitly, not via an opaque fallback function. |
| Security And Privacy | ❌ Error | PR introduces hard-coded home directory path `/home/leynos/.kani/...` and username in public documentation, replacing generic `$HOME` variable and dynamic version resolution. | Remove hard-coded absolute paths from docs/developers-guide.md line 490; restore dynamic path resolution using `$HOME` and version detection via `cargo kani --version`. |
| Rust Compiler Lint Integrity | ❌ Error | The `#[expect(clippy::missing_const_for_fn)]` on `is_available()` lacks a linked GitHub issue or tracked implementation item as required by the lint integrity check. | Add a GitHub issue reference to the `#[expect(...)]` reason, such as: `reason = "PyO3 FFI constraint; see issue` #NNN`"` or remove if not planned for future work. |
| Testing (Unit And Behavioural) | ⚠️ Warning | The PR adds unit tests that verify the probe helpers' classification, diagnostic naming, and fallback binding logic, plus a behavioural test using skipif. However, edge case testing is incomplete:... | Add unit tests parametrised over KeyboardInterrupt, SystemExit, and GeneratorExit to verify they are re-raised by `_crosshair_probe_failure_reason`, ensuring all exception paths are covered. |
| Architectural Complexity And Maintainability | ⚠️ Warning | The PR introduces five new abstractions for handling a single contingency (CrossHair availability): four helper functions, one exception class, and four dedicated test functions comprising ~99 line... | Flatten the helper layering: inline the probe classification logic into the module-level try-except block instead of delegating through \_is\_expected\_crosshair\_unavailable, \_crosshair\_unavailable\_reason, and \_crosshair\_probe\_failure\_reaso... |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat documentation and validation coverage as in scope).

### ❌ Failed checks (2 errors, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests lack integration coverage of the module-level probe. Tuple element order validation is absent—tests only check all values are None, missing incorrect ordering bugs. No test verifies actual ex... | Add integration test mocking module-level imports to verify probe exception handling; add assertions validating tuple element semantics (e.g., check that position 2 contains AnalysisKind, not another type). |
| Module-Level Documentation | ❌ Error | Module docstring does not document the new CrossHair capability probe mechanism that is central to the module's function since this PR. | Update the module docstring to explain the import-time CrossHair probe, expected failure handling, and fallback strategy for unavailable symbolic checks. |
| Observability | ⚠️ Warning | The PR introduces operational behaviour changes (runtime capability probing with graceful degradation) but lacks sufficient context in observability diagnostics. The warning message format only inc... | Enhance the warning diagnostic to include interpreter version, CrossHair availability status, and opcode compatibility details: emit Python version and CrossHair version (if available) alongside the exception type and message in the warn... |

**Comment by leynos** [comment]
@coderabbitai re:

> File: rust/cuprum-rust/src/lib.rs
> 
> Resolve the remaining `Rust Compiler Lint Integrity` failure.
> 
> Current state:
> - `is_available()` is annotated with:
>   `#[expect(clippy::missing_const_for_fn, reason = "...")]`
> - The reason string does not cite a tracked follow-up.
> - Repository issue search found no existing GitHub issue for this lint item.
> 
> Do one of the following, preferring the smallest correct change:
> 
> 1. If `is_available()` can legally and cleanly become `const` under PyO3, remove the `#[expect(...)]` and make the function comply with Clippy.
> 
> 2. Otherwise, keep the `#[expect(...)]` but first open a GitHub issue that tracks why this function must remain non-`const`, then update the attribute reason to reference that issue explicitly, for example:
>    `reason = "PyO3 runtime #[pyfunction] FFI boundary; see issue `#NNN`"`
> 
> Acceptance criteria:
> - No `#[expect(clippy::missing_const_for_fn)]` remains without a linked issue reference.
> - The reason string states the concrete technical constraint, not just that the function is intentionally non-const.
> - If the lint is retained, the linked GitHub issue is created and referenced in the code.
> - Keep the change tightly scoped to this lint failure.

It is not necessary to create an issue for every lint suppression. A tracking issue is only needed if the lint suppression is genuinely due to an issue to which a fix is planned. Where a suppression is needed due to the absence of a planned feature, the roadmap task in which the implementation is planned should be linked.

If the suppression is due to a bug in a third party library or the Rust language, and YOU know the tracking issue, feel free to provide it, but don't hassle people to raise an issue for something that doesn't need to be tracked.

**Comment by leynos** [comment]
@coderabbitai resume

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Excess Number of Function Arguments
cuprum/unittests/test_line_splitting.py: _CrossHairImportFailure.__call__

### What lead to degradation?
_CrossHairImportFailure.call has 5 arguments, max arguments = 4

### Why does this problem occur?
Functions with many arguments indicate either a) low cohesion where the function has too many responsibilities, or b) a missing abstraction that encapsulates those arguments.
The threshold for the Python language is 4 function arguments.

### How to fix it?
Start by investigating the responsibilities of the function. Make sure it doesn't do too many things, in which case it should be split into smaller and more cohesive functions. Consider the refactoring [INTRODUCE PARAMETER OBJECT](https://refactoring.com/catalog/introduceParameterObject.html) to encapsulate arguments that refer to the same logical concept.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Excess Number of Function Arguments
cuprum/unittests/test_line_splitting.py: _CrossHairImportFailure.__call__

### What lead to degradation?
_CrossHairImportFailure.call has 5 arguments, max arguments = 4

### Why does this problem occur?
Functions with many arguments indicate either a) low cohesion where the function has too many responsibilities, or b) a missing abstraction that encapsulates those arguments.
The threshold for the Python language is 4 function arguments.

### How to fix it?
Start by investigating the responsibilities of the function. Make sure it doesn't do too many things, in which case it should be split into smaller and more cohesive functions. Consider the refactoring [INTRODUCE PARAMETER OBJECT](https://refactoring.com/catalog/introduceParameterObject.html) to encapsulate arguments that refer to the same logical concept.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat documentation and validation coverage as in scope).

### ❌ Failed checks (1 error)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The new tests only cover CrossHair fallback/skip paths; they never assert the success path, so an always-skip probe would still pass. | Add a non-vacuous success-path test that imports the module with working CrossHair and asserts the bindings stay non-None and the contracts run. |

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai review
